### ISSUE TITLE
Feature request: Universal config

### ISSUE THREAD

**Issue by MaxGyver83** [david-janssen/kmonad#90]
Meanwhile I've got a lot of KMonad config files. And many of them are almost identical.

So two things could help:

1. In Linux, it would be great if I could pass the `device-file` as command line parameter. Then I wouldn't need one config per keyboard (laptop/internal and USB/external).
2. It would be easy to have a shared config for Windows, Linux and MacOS: On Windows, `device-file` and `iokit-name` could be ignored, and `input (low-level-hook)` and `output (send-event-sink)` could be set automatically. On MacOS, `device-file` could be ignored and `iokit-name` could be considered instead.

**Comment by thoelze1** [comment]
I like this idea but I should add that, unfortunately, the mac interface provides some OS-specific keycodes that (if present) would add complexity to a cross-compatible config

**Comment by MaxGyver83** [comment]
That's my workaround for my Linux configs (it's a fish function it should be easy to adapt it for bash):

`~/.config/fish/functions/k.fish`:

```fish
function k --description 'Start KMonad for all available keyboards'
    set config ~/.config/kmonad/vou4.kbd
    pkill kmonad

    echo "Activate KMonad for internal keyboard"
    kmonad $config & ; disown

    set keyb /dev/input/by-id/usb-Telink_Wireless_Receiver-if01-event-kbd
    if test -e $keyb
        echo "Activate KMonad for Jelly Comb keyboard"
        set dest "/tmp/kmonad-vou-jelly.kbd"
        cp $config $dest
        sed -i 's:/dev/input/by-path/platform-i8042-serio-0-event-kbd:'$keyb':' $dest
        kmonad $dest & ; disown
    end

    set keyb /dev/input/by-id/usb-SONiX_USB_DEVICE-event-kbd
    if test -e $keyb
        echo "Activate KMonad for Royal Kludge keyboard"
        set dest "/tmp/kmonad-vou-royal.kbd"
        cp $config $dest
        sed -i 's:/dev/input/by-path/platform-i8042-serio-0-event-kbd:'$keyb':' $dest
        kmonad $dest & ; disown
    end
end
```

**Comment by david-janssen** [comment]
Absolutely 100% agreed, and I opened a 'self-issue' to implement something like this at #62, although perhaps I should have been more verbose.

The solution I had in mind was the following:
1. All command-line options should have a corresponding entry in `defcfg`, so you can specify a default.
2. Any all `defcfg` entries should have command-line ways of specifying them, and these should always override `defcfg` if provided on the command line.

The second point you make about allowing more cross-platform keyboard-IO is also spot on. I think we can even expand it. You should be able to specify as many IO settings as you please, and *at least 1* has to work out for both input and output. Then you could also reuse the same config for different keyboards under the same OS. 

I'll close #62 now, because there is more pertinent information here.

**Comment by MaxGyver83** [comment]
Awesome! Sounds good!

**Comment by manna-harbour** [comment]
A command line option for `device-file` would be great, but how about also having a separate `defcfg` for each OS and auto-detecting which one to use, so the same file can be used on all platforms?

**Comment by theol0403** [comment]
+1 for command-line input configuration! 

It would be also useful for me to have something that allows the use of kmonad on multiple keyboards without the need for any shell scripting, I'm thinking something along the lines of having `device-file` accept a list of candidates where it hooks into the first one available. That way when I switch external keyboards it just works.

What would be even more awesome is having a multithreaded (or just self-spawning) kmonad which tries to hook into *all* available inputs. This way I could add my laptop keyboard and all external keyboards I use into the list, and they all just work without any shell scripting.

Finally, further down the road, it would be awesome if kmonad could wait for keyboards and be "hot-pluggable", meaning I could just run kmonad on OS boot, plug and unplug keyboards, and have kmonad continue to work on both my laptop and external boards all at once! This is all assuming I want the same configuration on all my internal and external ANSI keyboards.

However, after command-line configuration is supported, the rest could definitely be done just with clever shell scripting.

**Comment by MuhammedZakir** [comment]
> A command line option for `device-file` would be great, but how about also having a separate `defcfg` for each OS and auto-detecting which one to use, so the same file can be used on all platforms without modification?

Rather than auto-detecting, specify OS?

```
(defcfg
  "linux"
  ...
)

(defcfg
  "windows"
  ...
)

(defcfg
  "mac"
  ...
)
```

**Comment by manna-harbour** [comment]
@MuhammedZakir Yes, I meant a separate `defcfg` for each OS, specifying the OS, and have KMonad select the appropriate one based on the detected OS.  

Actually, this would also be useful in `deflayer` and `defalias` etc. as some keybindings depend on the OS.

**Comment by Swoogan** [comment]
Another thing to add to this request could be the ability to include other config files (like sourcing in bash, or git's `[include]` section). This way we could have one universal config that defines our layouts and includes a local config. The local config could then be used to add (and maybe override?) machine specific settings.

**Comment by MuhammedZakir** [comment]
> Another thing to add to this request could be the ability to include other config files (like sourcing in bash, or git's `[include]` section). This way we could have one universal config that defines our layouts and includes a local config. The local config could then be used to add (and maybe override?) machine specific settings.

Sorry, but I couldn't understand. What's the benefit of splitting a config? If you want to override `defcfg` part, you can do that using cli commands (currently available in `develop` branch).

**Comment by MaxGyver83** [comment]
If you just want to override a keyboard device file, a command line parameter is the best way to do it (IMHO). But if you want to add a layer only on Linux but not on Windows (or the other way round), then it's easier with a separate section or an included part in the config file.

**Comment by pataquets** [comment]
FYI: I've extracted the `include` feature into its own issue on #303.

**Comment by manna-harbour** [comment]
Here's a work around using a macro processor: #435.
